fgpyo/sam/__init__.py

```diff
@@ -1036,55 +1043,119 @@ class ReadEditInfo:
         op = elem.operator
-
-        if op == CigarOp.I:
-            insertions += 1
-            ins_bases += elem.length
-        elif op == CigarOp.D:
+        # TODO: use match-case statements after we drop Python 3.9 support
+        if op in (CigarOp.M, CigarOp.X, CigarOp.EQ):
+            for in_block_offset in range(0, elem.length):
+                if (target_offset + in_block_offset) >= len(reference_sequence):
+                    break  # out of bounds
+                query_base: str = rec.query_sequence[query_offset + in_block_offset].upper()
+                ref_base: str = reference_sequence[target_offset + in_block_offset].upper()
+                is_match = _is_match_base(
+                    query_base=query_base, ref_base=ref_base, match_htsjdk=match_htsjdk
+                )
+
+                if is_match:
+                    matches += 1
+                    current_match_count += 1
+                else:  # mismatch
+                    md_edits.append(str(current_match_count))  # append match count and reset
+                    current_match_count = 0
+                    md_edits.append(ref_base)  # grab mismatched base from the reference
+                    mms += 1
+            query_offset += elem.length_on_query
+            target_offset += elem.length_on_target
+        elif op == CigarOp.D:  # consumes ref
+            md_edits.append(str(current_match_count))  # append match count and reset
+            md_edits.append("^")
+            for in_block_offset in range(0, elem.length):
+                if (target_offset + in_block_offset) >= len(reference_sequence):
+                    break  # out of bounds
+            md_edits.append(f"{reference_sequence[target_offset:target_offset + elem.length]}")
+            current_match_count = 0
+            if target_offset < elem.length:
+                break
+            target_offset += elem.length
             deletions += 1
             del_bases += elem.length
-        elif op == CigarOp.M or op == CigarOp.X or op == CigarOp.EQ:
-            for i in range(0, elem.length):
-                q = rec.query_sequence[query_offset + i].upper()
-                t = reference_sequence[target_offset + i].upper()
-                if q != t or q not in ok_bases:
-                    mms += 1
-                else:
-                    matches += 1
-
-        query_offset += elem.length_on_query
-        target_offset += elem.length_on_target
+        elif op in (CigarOp.I, CigarOp.S):  # consumes query
+            query_offset += elem.length
+            if op == CigarOp.I:
+                insertions += 1
+                ins_bases += elem.length
+        elif op == CigarOp.N:  # skipped region from ref, consumes ref
+            target_offset += elem.length
+        elif op not in (CigarOp.H, CigarOp.P):  # pragma: no cover
+            raise ValueError(f"Invalid CIGAR operation: {op}")
+
+    md_edits.append(str(current_match_count))
```

Comment:
nit: f-strings are just a little faster
```suggestion
    md_edits.append(f"{current_match_count}")
```